libtransact/tests/state/merkle/sql_sqlite.rs

```diff
@@ -0,0 +1,147 @@
+// Copyright 2021 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! SQL-backed tests for the merkle state implementation
+
+use std::error::Error;
+
+use transact::state::merkle::sql::{
+    backend::{JournalMode, Sqlite, SqliteBuilder},
+    migration::MigrationManager,
+    SqlMerkleState,
+};
+use transact::{database::btree::BTreeDatabase, state::merkle::INDEXES};
+
+use super::*;
+
+fn new_sql_merkle_state_and_root(
+    dbpath: &str,
+) -> Result<(SqlMerkleState<Sqlite>, String), Box<dyn Error>> {
+    let backend = SqliteBuilder::new()
+        .with_connection_path(dbpath)
+        .with_journal_mode(JournalMode::Wal)
+        .with_create()
+        .build()?;
+    backend.run_migrations()?;
+
+    let state = SqlMerkleState::new(backend);
+    let initial_state_root_hash = state.initial_state_root_hash()?;
+
+    Ok((state, initial_state_root_hash))
+}
+
+#[test]
+fn merkle_trie_empty_changes() -> Result<(), Box<dyn Error>> {
+    run_test(|db_path| {
+        let (state, orig_root) = new_sql_merkle_state_and_root(db_path)?;
+        test_merkle_trie_empty_changes(orig_root, state);
+        Ok(())
+    })
+}
+
+#[test]
+fn merkle_trie_delete() -> Result<(), Box<dyn Error>> {
+    run_test(|db_path| {
+        let (state, orig_root) = new_sql_merkle_state_and_root(db_path)?;
+        test_merkle_trie_delete(orig_root, state);
+        Ok(())
+    })
+}
+
+#[test]
+fn merkle_trie_update() -> Result<(), Box<dyn Error>> {
+    run_test(|db_path| {
```

Comment:
This seems like it would be better done with a function macro specific to temporary-path wrapping/cleanup. (No idea if that's practical.)

Reply:
The current `run_test` matches the patterns used with the other file-based implementations (e.g. LMDB).  There are probably a couple of macro solutions that could be deployed.

A proc macro could make this very clean, though it would be a heavier lift from a development perspective (and probably beyond the scope of this PR):

For example:

```rust
#[test_with_temp_file_path]
fn merkle_trie_update(db_path: &str) -> Result<(), Box<dyn Error>> {
    ...
```

Advantages Could apply to all the tests with a file-backed DB implementation.
Disadvantage: Requires a separate crate just for the proc macro.

Alternatively, a standard macro could be used.  This would be easy  to implement, but would not quite look as much like a standard rust function.

For example:

```rust
test_with_temp_file!(merkle_trie_update, |db_path| {
    ...
})
```
Advantages: very easy to write, works with any file-backed DB implementation.
Disadvantage: syntax around the function seems a bit non-standard.

Reply:
Ok